in_toto/verifylib.py

```diff
@@ -634,6 +637,59 @@ def verify_all_item_rules(items, links, target_links=None):
     verify_item_rules(item.name, item.product_matchrules,
         link.products, target_links)
 
+def verify_threshold_equality(layout, step_link_dict):
+  """
+  <Purpose>
+    Iteratively verifies threshold conditions and link artifacts of
+    passed dictionary (step_link_dict).
+
+  <Arguments>
+    layout:
+            The layout specified by the project owner against which the
+            threshold will be verified.
+
+    step_link_dict:
+            A dictionary of key-link pair with step names as keys. For each
+            step name, there are one or more keyids and corresponding
+            link objects.
+
+  <Exceptions>
+    raises an Exception if threshold is not verified.
+    ThresholdVerificationError()
+
+  <Side Effects>
+    None.
+  
+  <Returns>
+    A dictionary containing one Link metadata object per step only if
+    the link artifacts of all link objects are identical for a step.
+  """
+  target_links = {}
+  for key_link in step_link_dict:
+    key_link_dict = step_link_dict[key_link]
+    
+    for step in layout.steps:
+       if step.name is key_link:
+        threshold = step.threshold
+
+    if len(key_link_dict) >= threshold:
+      link_dict = {}
+      if len(key_link_dict) == 1:
+        for keyid in key_link_dict:
+          link_dict = key_link_dict[keyid]
+      else:
+        i = 0
+        for keyid in key_link_dict:
+          if i == 0:
+            link_dict = key_link_dict[keyid]
+            i = i + 1
+          else:
+            if link_dict.materials != key_link_dict[keyid].materials or link_dict.products != key_link_dict[keyid].products:
+              raise ThresholdVerificationError("Link artifacts do not match!".format())
+      target_links[key_link] = link_dict
+    else:
+      raise ThresholdVerificationError("Step not performed by enough functionaries!".format())
```

Comment:
I think above algorithm can be simplified and I am not so sure if it does exactly what you intend to do. I am also not sure about what you intend to do, because there are not a lot of comments. :) Here is some general advise:
- Add comments that explain what you want to do
- Make sure that indentation is consistent
- Try to simplify your control flow statements, too much nesting makes it hard to read
- Read up on Python loops and how to use indices

And here is how I would write this function (in comments):
```python
# We are only interested in links that are related to steps defined in the
# Layout, so let's iterate over layout.steps
 
  # Skip steps that don't require multiple functionaries
  ...

  # Extract the step_link_dict for this step from the passed chain_link_dict
  link = chain_link_dict.get(step.name)
  ...

  # Check if we have at least <threshold> links for this step
  ...

  # AFICS `verify_all_steps_signatures` only verifies if each link of all steps
  # was signed by ANY of the functionaries authorized to perform this step.
  # But we also have to make sure we have at least <threshold> UNIQUE links,
  # i.e. signed by different functionaries. We should check this here and we
  # should not use the functionary keyid used as dictionary key, but
  # rather the keyid in the link's signature, because the former is taken from
  # the file name and could be forged.
  ...

  # Finally take one exemplary link (e.g. the first in the step_link_dict) and
  # iterate over the remaining links to compare their properties, above all
  # materials and products have to be equal for all links of a step.
  # I'm not sure if we care about byproducts, command and return value
  # maybe we should discuss this.
  
  # If no exception was raised so far, everything should be fine (the pythonic way)
```